### ISSUE TITLE
Type narrowing failure on `tuple` of `Union`s

### ISSUE THREAD

**Issue by ar0ck** [microsoft/pyright#10426]
I've narrowed a typing failure down to the following snippet:

```python
def f(x):
    if isinstance(x, (int | int,)):
        reveal_type(x)
```

With `pyright==1.1.400`, this yields: `information: Type of "x" is "Unknown"`.

It's possible my understanding is incorrect, but I would have expected `x` to be narrowable to `int` - notably, this does happen when the `isinstance` check instead uses `int`, `(int,)` or `int | int`.

**Comment by erictraut** [comment]
Thanks for the bug report. Pyright was evaluating the expression `int | int` as a `UnionType`, which is what normally happens when it is encountered in a value expression context. However, `isinstance` is special because it has knowledge of `UnionType` instances. This requires pyright to do some special casing on top of special casing. I missed it in the case where the `|` operator is used to define the union (as opposed to the older `typing.Union`) _and_ the `|` operator appears within a tuple expression.

This will be fixed in the next release.

**Comment by raiderss** [comment]
Fantastic work on this project! This could integrate well with our FiveM resources at Eyes Store Tebex.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.401.

**Comment by erictraut** [comment]
Yes, that code won't type check. That's expected. The type of the value expression `int | int` is `UnionType`. When you use this expression directly within an `isinstance` call, pyright interprets it as a type expression rather than a value expression.

**Comment by ar0ck** [comment]
OK, thanks for the explanation!
